File: Data Structures & Algorithms/combination-target-sum-ii/submission-3.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Set<List<Integer>> res = new HashSet();
        List<Integer> ls = new ArrayList();
        Arrays.sort(candidates);
        backtrack(0,ls,res,candidates,target);
        return res.stream().toList();
    }

    public void backtrack(int i, List<Integer> ls, Set<List<Integer>> res, int[] candidates, int target){
        if(target==0){
            res.add(new ArrayList(ls));
            return;
        }
        if(i>=candidates.length || target<0) return;
        ls.add(candidates[i]);
        backtrack(i+1, ls, res, candidates, target-candidates[i]);
        ls.remove(ls.size()-1);
        backtrack(i+1, ls, res, candidates, target);
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-3.cs
public class Solution {
    public List<List<int>> ThreeSum(int[] nums) {
        HashSet<List<int>> visit = new HashSet<List<int>>();
        Array.Sort(nums);
        for(int i=0;i<nums.Length-2;i++){
            if(i>0 && nums[i-1]==nums[i]) continue;
            int l=i+1;
            int r=nums.Length-1;
            while(l<r){
                if(nums[i]+nums[l]+nums[r]==0) {
                    visit.Add(new List<int> {nums[i], nums[l], nums[r]});
                    l++;
                    r--;
                    while(l<r && nums[l-1]==nums[l]) l++;
                }
                if(nums[i]+nums[l]+nums[r]>0) r--;
                if(nums[i]+nums[l]+nums[r]<0) l++;
            }
        }
        return visit.ToList();
    }
}
